lib/active_model/serializer.rb

```diff
@@ -108,20 +110,32 @@ def self.digest_caller_file(caller_line)
       Digest::MD5.hexdigest(serializer_file_contents)
     end
 
-    def self.get_serializer_for(klass)
+    def self.get_serializer_for(klass, parent_serializer, controller)
       serializers_cache.fetch_or_store(klass) do
         serializer_class_name = "#{klass.name}Serializer"
-        serializer_class = serializer_class_name.safe_constantize
+
+        nested_serializer_class_name = "#{self}::#{serializer_class_name}"
+        serializer_class = nested_serializer_class_name.safe_constantize
+
+        if parent_serializer
+          namespaced_serializer_class_name = "#{parent_serializer.root_serializer.class.name.deconstantize}::#{serializer_class_name}"
+          serializer_class ||= namespaced_serializer_class_name.safe_constantize
+        elsif controller
+          controller_namespaced_serializer_class_name = "#{controller.class.name.deconstantize}::#{serializer_class_name}"
+          serializer_class ||= controller_namespaced_serializer_class_name.safe_constantize
```

Comment:
That's why I use `||=`. What this piece does is look for a `RelatedSerializer` serializer in:
 + `CurrentSerializer::RelatedSerializer` (which would be `ActiveModel::Serializer::RelatedSerializer` if it is not actually a related serializer, but that would fail so it's ok)
 + revert to `My::Name::Space::RelatedSerializer` if called from `My::Name::Space::CurrentSerializer`
 + or revert to `My::Controller::Name::Space::RelatedSerializer` if called from `My::Controller::Name::Space::SomeController`
 + finally, revert to `::RelatedSerializer` if none of the above exist